state/cluster/badger/mutator.go

```diff
@@ -36,202 +37,286 @@ func NewMutableState(state *State, tracer module.Tracer, headers storage.Headers
+// a candidate extending cluster block.
+// No errors are expected during normal operation.
+func (m *MutableState) getExtendCtx(candidate *cluster.Block) (extendContext, error) {
+	var ctx extendContext
+	ctx.candidate = candidate
+
+	err := m.State.db.View(func(tx *badger.Txn) error {
+		// get the latest finalized cluster block and latest finalized consensus height
+		err := procedure.RetrieveLatestFinalizedClusterHeader(candidate.Header.ChainID, &ctx.finalizedClusterBlock)(tx)
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized cluster head: %w", err)
+		}
+		err = operation.RetrieveFinalizedHeight(&ctx.finalizedConsensusHeight)(tx)
+		if err != nil {
+			return fmt.Errorf("could not retrieve finalized height on consensus chain: %w", err)
+		}
+
+		err = operation.RetrieveEpochFirstHeight(m.State.epoch, &ctx.epochFirstHeight)(tx)
+		if err != nil {
+			return fmt.Errorf("could not get operating epoch first height: %w", err)
+		}
+		err = operation.RetrieveEpochLastHeight(m.State.epoch, &ctx.epochLastHeight)(tx)
+		if err != nil {
+			if errors.Is(err, storage.ErrNotFound) {
+				ctx.epochHasEnded = false
+				return nil
+			}
+			return fmt.Errorf("unexpected failure to retrieve final height of operating epoch: %w", err)
+		}
+		ctx.epochHasEnded = true
+		return nil
+	})
+	if err != nil {
+		return extendContext{}, fmt.Errorf("could not read required state information for Extend checks: %w", err)
+	}
+	return ctx, nil
+}
+
 // Extend validates that the given cluster block passes compliance rules, then inserts
 // it to the cluster state.
 // TODO (Ramtin) pass context here
 // Expected errors during normal operations:
 //   - state.OutdatedExtensionError if the candidate block is outdated (e.g. orphaned)
+//   - state.UnverifiableExtensionError if the candidate block cannot be verified
 //   - state.InvalidExtensionError if the candidate block is invalid
 func (m *MutableState) Extend(block *cluster.Block) error {
-
 	blockID := block.ID()
+	header := block.Header
+	payload := block.Payload
 
 	span, ctx := m.tracer.StartCollectionSpan(context.Background(), blockID, trace.COLClusterStateMutatorExtend)
 	defer span.End()
 
-	err := m.State.db.View(func(tx *badger.Txn) error {
+	setupSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendSetup)
+	// check chain ID
+	if header.ChainID != m.State.clusterID {
+		return state.NewInvalidExtensionErrorf("new block chain ID (%s) does not match configured (%s)", block.Header.ChainID, m.State.clusterID)
+	}
 
-		setupSpan, _ := m.tracer.StartSpanFromContext(ctx, trace.COLClusterStateMutatorExtendSetup)
+	// check for a specified reference block
+	// we also implicitly check this later, but can fail fast here
+	if payload.ReferenceBlockID == flow.ZeroID {
+		return state.NewInvalidExtensionError("new block has empty reference block ID")
+	}
 
-		header := block.Header
-		payload := block.Payload
+	// get the header of the parent of the new block
```

Comment:
I am missing documentation in the interface and the implementation clearly stating that we expect the parent block to be known. 

* I would be inclined to add a dedicated sentinel error (like 'UnknownParentError'). 
* The caller can then decide whether they expect this error or not. 

alternatively, I am totally fine with just documenting in the `MutableState` interface and here that the parent block _must_ be known (resulting in an exception otherwise)